src/types.rs

```diff
@@ -422,13 +446,75 @@ impl<N: Name> Type<N> {
             }
             Type::Variable(v) => {
                 *self = ctx
-                    .substitution
+                    .cache
                     .get(&v)
+                    .or_else(|| ctx.substitution.get(&v))
                     .map(|tp| tp.apply(ctx))
                     .unwrap_or_else(|| Type::Variable(v));
             }
         }
     }
+    /// Applies the type in a [`Context`], performing path compression.
+    /// 
+    /// Path compression is a technique commonly used in [Union-Find data structures]. We apply it
+    /// here so that whenever a chain of substitutions is traversed, each variable is updated to
+    /// point to its ultimate value. For example, the chain:
+    ///
+    /// `t0 ↦ t1`, `t1 ↦ t2`, and `t2 ↦ int`
+    ///
+    /// becomes
+    /// 
+    /// `t0 ↦ int`, `t1 ↦ int`, and `t2 ↦ int`
+    ///
+    /// Rather than updating the actual mappings, [`Context`] maintains a `cache` of these
+    /// compressed mappings.
+    ///
+    /// [Union-Find data structure]: https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+    /// [`Context`]: struct.Context.html
+    pub fn apply_compress(&self, ctx: &mut Context<N>) -> Type<N> {
+        match *self {
+            Type::Constructed(ref name, ref args) => {
+                let args = args.iter().map(|t| t.apply_compress(ctx)).collect();
+                Type::Constructed(name.clone(), args)
+            }
+            Type::Variable(v) => ctx
+                .cache
+                .get(&v)
+                .or_else(|| ctx.substitution.get(&v))
+                .cloned()
+                .map(|mut tp| {
+                    tp.apply_mut_compress(ctx);
+                    ctx.cache.insert(v, tp.clone());
```

Comment:
This clones `tp` even if there’s no change to `ctx.cache`, it might be better to wrap this in a check `if cache.get(&v) != Some(&tp)` to avoid cloning.

Reply:
Sure - no problem.